### PR TITLE
Handle stride > 1 with im2col in CUDA thnn conv2d

### LINKED ISSUE #54036: Silently incorrect convolution on CUDA without cuDNN
## 🐛 Bug

```
import torch
import torch.nn as nn
model = nn.Conv2d(64, 128, [1, 1], [2, 2], [0, 0])
x = torch.randn(1, 64, 8, 8)

model.cuda()
result_cudnn = model(x.cuda())

torch.backends.cudnn.enabled=False
result = model(x.cuda())

assert torch.allclose(result, result_cudnn)
```
The above code asserts. This convolution (with batch size > 1) is important because it's present in resnet; the bug makes it so that resnet stops training. I guess I am the first person to run into this so the conclusion may be that no one uses CUDA convolutions without cudnn.

## Additional context

On master, but I can provide more information as necessary.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411

### PR DESCRIPTION
**Pull request by t-vi**
The fallback thnn 2d convolution uses `im2col` to get patches and `gemm` to implement convolution .
I has a shortcut to use `gemm` directly for kernel size 1, but this only works for stride == 1 and padding == 0. 
This PR adds checks for stride == 1 and padding == 0 to determining whether `im2col` can be skipped.

Fixes #54036

### CODE REVIEW

**Comment by t-vi** [comment]
@zou3519 would you be willing to look at the fix?

**Comment by zou3519** [comment]
Sure, I can take a look. I'll either review it or find someone else if I don't know what's going on :)

**Comment by t-vi** [comment]
While investigating this further, it occurred to me that padding > 0 and kernel size = 1 was also not properly checked and we need to fallback there, too. So I added that.

**Comment by zou3519** [comment]
@ngimel could you take a look at this as well? This looks correct to me but I'm not familiar with how convolutions work. 

#44600 was the PR that added this behavior -- do we also need to change the native convolution transpose code? (If not, could we add some tests to confirm?)

**Comment by ngimel** [comment]
Why is padding >= kernel_size accepted as a valid convolution arg? I though padding is supposed to be < kernel_size, but I might be mistaken.

**Comment by t-vi** [comment]
I'd imagine that the padding limitation is a CuDNN restriction mostly. I don't have much of an opinion whether we might return error instead, but my main goal is not to give a wrong result.

**Comment by ngimel** [comment]
Totally agree! I've checked, and we do accept padding >=kernel_size, and apparently cudnn accepts it too, at least it doesn't error out.

File: aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp [hunk 1]
```diff
@@ -453,7 +453,9 @@ static void slow_conv_transpose2d_backward_out_cpu_template(
           grad_input_n = grad_input.select(0, elt);
           grad_output_n = grad_output.select(0, elt);
 
-          if (kernel_height != 1 || kernel_width != 1) {
+          if (kernel_height != 1 || kernel_width != 1 || stride_height != 1 ||
```
Comment (ngimel) [inline aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp:456]: Given how many times these checkes are repeated, does it make sense to have a helper function somewhere in ConvUtils.h?

**Review by ngimel — APPROVED** [review]
The fix and the test looks good, thanks @t-vi!

File: aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp (hunk 1 above)
Comment (t-vi) [reply aten/src/ATen/native/NaiveConvolutionTranspose2d.cpp:456]: So this would replace a lot of ` != 1 || ` with `, `. It's shorter, but not by that much if we enumerate the parameters.

**Comment by zou3519** [comment]
Test failures look unrelated, will merge if/when internal CI passes

**Review by zou3519 — APPROVED** [review]
